packages/@expo/cli/src/export/exportAssets.ts

```diff
@@ -42,17 +73,35 @@ export async function resolveAssetBundlePatternsAsync<T extends ExpoConfig>(
       const shouldBundle = shouldBundleAsset(asset, fullPatterns);
       if (shouldBundle) {
         debug(`${shouldBundle ? 'Include' : 'Exclude'} asset ${asset.files?.[0]}`);
-        return asset.fileHashes.map(
-          (hash) => 'asset_' + hash + ('type' in asset && asset.type ? '.' + asset.type : '')
-        );
+        return asset.fileHashes.map((hash) => mapAssetHashToAssetString(asset, hash));
       }
       return [];
     })
     .flat();
 
   // The assets returned by the RN packager has duplicates so make sure we
   // only bundle each once.
-  (exp as any).bundledAssets = [...new Set(allBundledAssets)];
+  return new Set(allBundledAssets);
+}
+
+/**
+ * Resolves the assetBundlePatterns from the manifest and returns a list of assets to bundle.
+ *
+ * @modifies {exp}
+ */
+export async function resolveAssetPatternsToBeBundledAsync<T extends ExpoConfig>(
+  projectRoot: string,
+  exp: T,
+  assets: Asset[]
+): Promise<T & { bundledAssets?: Set<string> }> {
```

Comment:
instead of returning the modified exp, we should return void to make it clearer that this mutates in-place.

Then no need for generic.